### PR TITLE
Steal the algorithm used to combine hashes from tupleobject.c

### PR DESCRIPTION
**Pull request by mikegraham**
closes #15224

### CODE REVIEW

**Comment by mikegraham** [comment]
Here's an initial pass at stealing https://github.com/python-git/python/blob/master/Objects/tupleobject.c#L290 for the combining. I am not 100% that the problem is my (rather crude) combiner, but possibly the exact way we're using the bitmixer in hash_array. I'm trying to think about it........I think we might be maintaining undesirable linearity.

May I ask, how did you encounter these collisions?

**Comment by mikegraham** [comment]
If the basic approach looks sound I can add some comments around some of the lazy iterator wackiness.

File: pandas/tools/hashing.py [hunk 1]
```diff
@@ -13,6 +14,20 @@
 _default_hash_key = '0123456789123456'
 
 
+def _combine_hash_arrays(arrays, num_items):
+    first = next(arrays)
+    arrays = itertools.chain([first], arrays)
+
+    mult = np.zeros_like(first) + np.uint64(1000003)
+    out = np.zeros_like(first) + np.uint64(0x345678L)
```
Comment (jreback) [inline pandas/tools/hashing.py]: the L is not working in py3. (remove it and its ok).

**Comment by jreback** [comment]
@mikegraham the collisions I found by hashing 

```
In [3]: i = pd.MultiIndex.from_product([np.arange(1000),np.arange(1000)],names=['one','two'])

In [4]: i.to_dataframe(index=False) 
```

which is basically a cartesian product of 1000 x 1000. nothing special really, just a test case I am using.

**Comment by jreback** [comment]
closing in favor of in #15224 

thanks @mikegraham

**Pull request closed without merging**